### ISSUE TITLE
Mis-compilation of infinite recursions

### ISSUE THREAD

**Issue by ksqsf** [rust-lang/rust#85742]
Consider [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fb991afd95d7ece22a6711d70a26e3b3)

```rust
use std::time::Instant;

fn collatz(i: usize) -> bool {
    if i == 1 {
        true
    } else if i % 2 == 0 {
        collatz(i / 2)
    } else {
        collatz(3 * i + 1)
    }
}

fn main() {
    static TIMES: usize = 500000;
    let start = Instant::now();
    for i in 0..TIMES {
        if ! collatz(i) {
            println!("{}", i);
            return;
        }
    }
    println!("Done, {} millis elapsed!", start.elapsed().as_millis());
}
```

It contains infinite recursion (`collatz(0)`), but the whole loop gets optimized away with Nightly + Release.

Related: #28728 (Perhaps also related to [this commit](https://github.com/rust-lang/rust/commit/0517acd54353869b2fbfc50af61ea7bd1fd309e0)).

Not sure if this is an LLVM bug?

**Comment by moxian** [comment]
This requires `--release` to elide the function, but is reproducible on both nightly and stable (and as far back as at least 1.15.0) (so technically not a regression)

**Comment by oxalica** [comment]
In addition:
1. According to #85750, using `0..=(TIMES - 1)` instead of `0..TIMES`  produces the correct result (infinite loop).
2. Decrease `TIMES` to less than 100, both `..` and `..=` get mis-compiled to an immediate return.
   Maybe related to loop-unroll or inline?

**Comment by nikic** [comment]
Upstream issue: https://bugs.llvm.org/show_bug.cgi?id=50511

**Comment by apiraino** [comment]
thanks @nikic and @oxalica for the analysis.

Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2385742.20Mis-compilation.20of.20infinite.20recursions) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-critical

**Comment by nikic** [comment]
Upstream fix: https://github.com/llvm/llvm-project/commit/1b84acb23acac2fbb450312049495164a16ee715

**Comment by nikic** [comment]
This issue has been fixed a while ago by #87018, which picks up the referenced fix.
